packages/liquidator/src/liquidationStrategy.js

```diff
@@ -0,0 +1,223 @@
+    assert(currentBlockTime >= 0, "requires currentBlockTime");
+    assert(maxCollateralPerToken, "requires maxCollateralPerToken");
+    return [
+      sponsor,
+      { rawValue: liquidationMinPrice.toString() },
+      { rawValue: maxCollateralPerToken.toString() },
+      { rawValue: tokensToLiquidate.toString() },
+      parseInt(currentBlockTime) + liquidationDeadline
+    ];
+  }
+
+  function min(a, b) {
+    return toBN(a).lte(toBN(b)) ? a : b;
+  }
+  function max(a, b) {
+    return toBN(a).gte(toBN(b)) ? a : b;
+  }
+
+  // use in the case of a non whale defense liquidation
+  // Similar to the logic that already exists, but takes into account more information
+  function calculateTokensToLiquidate({
+    syntheticTokenBalance,
+    positionTokens,
+    empMinSponsorSize,
+    maxTokensToLiquidateWei
+  }) {
+    positionTokens = toBN(positionTokens);
+    empMinSponsorSize = toBN(empMinSponsorSize);
+    syntheticTokenBalance = toBN(syntheticTokenBalance);
+
+    if (positionTokens.lte(0)) return toBN(0);
+
+    // available capital calculates the max we can currently spend
+    let availableCapital = syntheticTokenBalance.sub(toBN(whaleDefenseFundWei));
+    if (maxTokensToLiquidateWei) availableCapital = min(toBN(maxTokensToLiquidateWei), availableCapital);
+
+    // we have enough capital to fully liquidate positon
+    if (availableCapital.gte(positionTokens)) {
+      return positionTokens;
+    }
+
+    // we dont have enough capital for full liquidation.
+    // we check to make sure position remains above min sponsor size
+    if (positionTokens.sub(availableCapital).gt(empMinSponsorSize)) {
+      return availableCapital;
+    }
+
+    // our liquidation would bring it below min sponsor size, so just bring it to min
+    return max("0", positionTokens.sub(empMinSponsorSize));
+  }
+
+  // Call this with only undercollateralized position
+  // SyntheticTokenBalance should be passed in respecting maxTokensToLiquidate
+  function processPosition({
+    position,
+    syntheticTokenBalance,
+    currentBlockTime,
+    maxCollateralPerToken,
+    // mandated token to liquidate amount, an override
+    maxTokensToLiquidateWei,
+    empMinSponsorSize,
+    ...logInfo
+  }) {
+    assert(empMinSponsorSize, "requires empMinSponsorSize");
+    assert(syntheticTokenBalance, "requires syntheticTokenBalance");
+    assert(currentBlockTime >= 0, "requires currentBlockTime");
+    assert(position, "requires position");
+
+    empMinSponsorSize = toBN(empMinSponsorSize);
+
+    // whale fund enabled
+    if (whaleDefenseFundWei && toBN(whaleDefenseFundWei).gt(toBN("0"))) {
+      // we should only liquidate the minimum here to extend withdraw
+      if (
+        shouldLiquidateMinimum({
+          position,
+          syntheticTokenBalance,
+          empMinSponsorSize,
+          currentBlockTime
+        })
+      ) {
+        // we are using the WDF to liquidate. We need to alert user
+        emit("log", "info", {
```

Comment:
This is an interesting pattern: you are injecting an `emit` function into the `liquidationStrategy.js` and then spreading the output in a Winston log within the `Liquidator`. 

However, this pattern differs from how we normally do logging where we would inject a `logger` instance into the class/library/package and then re-use the logger all the way down. Can you explain a bit around your rationale for changing this pattern here?

Offline we've discussed changing all logs to work like this and handling them at the top layer of the bots. I wanted to ask if we think it's ok to have this function dealing with logging differently? To be clear I do like this pattern but it feels a bit different from how we do logging in other places and want to be sure we are ok With this inconsistency.

Reply:
I purposefully wanted to show a different way to do logging here. I took out the example in the mocha tests, but I was listening to an event emitter to check that logs were emitted. I think in this next pass I will add that back in so you can see how this compares to the existing way we test logs. 

There are 2 main advantages I see with this method:
- no dependency on the logger so testing can really be much easier and cleaner when testing for logs, or any other events for that matter. Dont need to mock the logger or have all this tooling around getting visibility into it. 
-  you can get slightly more code re-use. You can imagine having a logging function (initialized in constructor) which attaches all the default data for this class, such as the `at` or any configuration variables. This way logs are consistent and additional context never needs to be passed in. Next pass will show this in a more generalized case. 

You can also imagine not even injecting logger in liquidator at all, and emitting events all the way up to the index file to do the logging. Not something I want to refactor, but its just here to show an alternate way.

Reply:
For sure. I think it's a great pattern. I'm happy to leave it like this for now in this section and we can incorporate it into other parts of the bots later to break down this dependency in the tests.